### PR TITLE
Enable CD as sidecar with vanilla pulumi-nodejs image

### PR DESCRIPTION
**Pull request by edwardrf**

### CODE REVIEW

File: src/pkg/aws/ecs/cfn/template.go [hunk 1]
```diff
@@ -283,35 +289,93 @@ func createTemplate(stack, image string, memory float64, vcpu float64, spot bool
...
+					"awslogs-group":         cloudformation.Ref(_logGroup),
+					"awslogs-region":        cloudformation.Ref("AWS::Region"),
+					"awslogs-stream-prefix": awsecs.AwsLogsStreamPrefix,
+				},
+			},
+			VolumesFrom:      volumesFrom,
+			MountPoints:      mountPoints,
+			EntryPoint:       task.EntryPoint,
+			Command:          task.Command,
+			WorkingDirectory: ptr.String("/app"), // TODO: make this configurable
```
Comment (lionello) [inline src/pkg/aws/ecs/cfn/template.go:361]: add to `Task`?

**Review by lionello — CHANGES_REQUESTED** [review]

File: src/pkg/cli/client/byoc.go [hunk 2]
```diff
@@ -42,7 +43,7 @@ const (
 
 var (
 	// Changing this will cause issues if two clients with different versions are using the same account
-	cdImage = pkg.Getenv("DEFANG_CD_IMAGE", "public.ecr.aws/defang-io/cd:beta")
+	cdImage = pkg.Getenv("DEFANG_CD_IMAGE", "public.ecr.aws/defang-io/cd:public-beta")
```
Comment (lionello) [inline src/pkg/cli/client/byoc.go:46]: ?? we haven't changed the GitHub Action that copies the image there. Also, that repo should only ever have public images, so the extra "public-" seems superfluous.

File: src/pkg/aws/ecs/cfn/template.go (hunk 1 above)
Comment (edwardrf) [reply src/pkg/aws/ecs/cfn/template.go]: Added both WorkDir and DependsOn to the Container (previously task) struct

File: src/pkg/cli/client/byoc.go (hunk 2 above)
Comment (edwardrf) [reply src/pkg/cli/client/byoc.go:46]: It is mainly to break the dependency loop between cli and different CD version, preventing older cli version from using the new CD image which will be missing pulumi, or the new cli using the older CD image which does not expose the right volumes.

File: src/pkg/docker/run_test.go [hunk 3]
```diff
@@ -16,7 +16,7 @@ func TestRun(t *testing.T) {
 
 	d := New()
 
-	err := d.SetUp(context.TODO(), []types.Container{{Image: "alpine:latest", Platform: d.platform}})
+	err := d.SetUp(context.TODO(), []types.Container{{Image: "alpine:latest", Memory: 6 * 1024 * 1024, Platform: d.platform}})
```
Comment (lionello) [inline src/pkg/docker/run_test.go:19]: shouldn't be needed, since our "fixup" logic would pick the smallest

File: src/pkg/types/driver.go [hunk 4]
```diff
@@ -21,6 +21,8 @@ type Container struct {
 	VolumesFrom []string // container (default rw), container:rw, or container:ro
 	EntryPoint  []string
 	Command     []string // overridden by Run()
+	WorkDir     *string
```
Comment (lionello) [inline src/pkg/types/driver.go:24]: why ptr? `""` is a good default here, no?

File: src/pkg/types/driver.go [hunk 5]
```diff
@@ -21,6 +21,8 @@ type Container struct {
...
+	DependsOn   map[string]string // container name -> condition
```
Comment (lionello) [inline src/pkg/types/driver.go:25]: for now, could we simply detect dependencies based on the volumes?

File: src/pkg/cli/client/byoc.go [hunk 6]
```diff
@@ -118,13 +119,38 @@ func (b *byocAws) setUp(ctx context.Context) error {
...
-		Name:     awsecs.ContainerName,
-		Memory:   512_000_000, // 512 MB
-		Platform: "linux/amd64",
-	}}
+	cdTaskName := cdTaskPrefix
+	containers := []types.Container{
+		{
+			Image:     "pulumi/pulumi:latest",
+			Name:      awsecs.ContainerName,
+			Memory:    4 * 512_000_000, // 512 MiB
```
Comment (lionello) [inline src/pkg/cli/client/byoc.go:127]: change or remove comment

File: src/pkg/cli/client/byoc.go [hunk 7]
```diff
@@ -118,13 +119,38 @@ func (b *byocAws) setUp(ctx context.Context) error {
...
-	containers := []types.Container{{
-		Image:    cdImage,
-		Name:     awsecs.ContainerName,
-		Memory:   512_000_000, // 512 MB
-		Platform: "linux/amd64",
-	}}
+	cdTaskName := cdTaskPrefix
+	containers := []types.Container{
+		{
+			Image:     "pulumi/pulumi:latest",
```
Comment (lionello) [inline src/pkg/cli/client/byoc.go:125]: No registry defaults to Docker. Let's use the one from public ecr.

**Review by lionello — CHANGES_REQUESTED** [review]

File: src/pkg/types/driver.go (hunk 4 above)
Comment (edwardrf) [reply src/pkg/types/driver.go:24]: So we can pass in nil to allow the default `WORKDIR` specified in the dockerfile to take effect, setting it to "" may modify the behaviour.

Comment (lionello) [reply src/pkg/types/driver.go:24]: just detect `""` and pass `nil`? `WorkDir: ""` makes no sense anyway

File: src/pkg/types/driver.go (hunk 5 above)
Comment (lionello) [reply src/pkg/types/driver.go:25]: then, lets at least make an enum out of the condition string.

File: src/pkg/types/driver.go (hunk 4 above)
Comment (edwardrf) [reply src/pkg/types/driver.go:24]: I think using pointer probably is more idiomatic to represent an optional type due to how aws sdk use it nowadays.

File: src/pkg/types/driver.go (hunk 5 above)
Comment (edwardrf) [reply src/pkg/types/driver.go]: Good point, added.

File: src/pkg/aws/ecs/cfn/template.go [hunk 8]
```diff
@@ -373,16 +384,18 @@ func createTemplate(stack string, containers []types.Container, spot bool) *clou
...
+			EntryPoint:       container.EntryPoint,
+			Command:          container.Command,
+			WorkingDirectory: container.WorkDir,
+			DependsOnProp:    dependsOn,
 		}
 		containerDefinitions = append(containerDefinitions, def)
 	}
 
 	const _taskDefinition = "TaskDefinition"
+	const cdTaskDefName = awsecs.ContainerName + "-cd"
```
Comment (lionello) [inline src/pkg/aws/ecs/cfn/template.go:398]: unused

File: src/pkg/aws/common.go [hunk 9]
```diff
@@ -23,9 +23,10 @@ func (r Region) String() string {
 
 func (a *Aws) LoadConfig(ctx context.Context) (aws.Config, error) {
 	cfg, err := LoadDefaultConfig(ctx, a.Region)
-	if err == nil {
-		a.Region = Region(cfg.Region)
+	if err != nil {
+		return cfg, err
 	}
+	a.Region = Region(cfg.Region)
 	if a.Region == "" {
```
Comment (lionello) [inline src/pkg/aws/common.go]: check `cfg.Region` and mutate `a.Region` after the `if`

File: src/pkg/cli/client/byoc.go [hunk 10]
```diff
@@ -118,13 +119,39 @@ func (b *byocAws) setUp(ctx context.Context) error {
...
-		Memory:   512_000_000, // 512 MB
-		Platform: "linux/amd64",
-	}}
+	cdTaskName := cdTaskPrefix
+	containers := []types.Container{
+		{
+			Image:     "public.ecr.aws/pulumi/pulumi-nodejs:latest",
+			Name:      awsecs.ContainerName,
+			Cpus:      0.5,
+			Memory:    2048_000_000, // 2G
```
Comment (lionello) [inline src/pkg/cli/client/byoc.go:128]: add a `_` between 2 and 048

**Review by lionello — APPROVED** [review]

**Pull request merged**